### DISCUSSION TITLE
Is it possible to import already existing DXF files in a DXF file created with EZDXF?

### DISCUSSION THREAD

**Discussion by thankyourobert** [mozman/ezdxf#825] [Q&A]
Hello all 

First of all many thanks for the great library! 

I'm looking into being able to "add" a DXF file into a DXF file created with the EZDXF library. 
I have written a simple script who's making a rectangle with 2 circles: [test_import.txt](https://github.com/mozman/ezdxf/files/10530468/test_import.txt) 

```Python
import pathlib
import datetime
import numpy as np
import pandas as pd
import os
import ezdxf
from math import *
filename = pathlib.Path(__file__).parent.absolute().__str__() 
desktop_filename = os.path.join(os.path.join(os.environ['USERPROFILE']), 'Desktop') 
updated_directory_filename = os.path.join(desktop_filename, "Output_updated")
templates_filename = pathlib.Path(__file__).parent.absolute().__str__() +'/Templates/'
database_filename = os.path.join(filename, "Database")
data_instruments_filename = os.path.join(database_filename, "Data_Instruments.xlsx")
new_directory_filename = os.path.join(desktop_filename, "Output")


lengte_rechthoek = 120

straal = 4.9*0.5

doc = ezdxf.new('R2018')# create a new DXF R2010 drawing, official DXF version name: 'AC1024'
doc.styles.new("STD",dxfattribs= {'font': 'Arial'})
msp = doc.modelspace()  # add new entities to the modelspace
i =1
j=i
points = [(1+i*123, 1+j*33), (i*123+121, 1+j*33), (i*123+121, 31+33*j), (1+i*123, 31+33*j),(1+i*123, 1+j*33)]
msp.add_lwpolyline(points,dxfattribs={'color': 1})
msp.add_circle(((1+6.45+123*i), 1.5+(23.25-5-3.75)+j*33),dxfattribs={'color': 1},radius = straal)
msp.add_circle(((1+(lengte_rechthoek-6.45)+123*i),1.5+(23.25-5-3.75)+j*33),dxfattribs={'color': 1},radius = straal)
#is there a function that is having a 1) dxl file & 2) x-y position  as arguments? 
msp.add_dxf_file("my_dxf_file.dxf",x=100,y=50)

file_created = filename+"/DXF_Files/"+str(datetime.date.today())+'_DXF_TAG_plates'+'.dxf'
doc.saveas(file_created)
```

Is there a function that would be able to take a DXF file as an argument and to be able to position him in another DXF file? 
![image](https://user-images.githubusercontent.com/83846226/215346289-70644dab-15d7-4022-b487-ab66fbf671c1.png)

Example below: 
This is the output of the script: 

![image](https://user-images.githubusercontent.com/83846226/215345994-a740a265-0e8c-46d7-a69e-c73b95e53724.png)

With the function described above I want to be able to place the lasercutting logo where I want in another DXF- file. 

![image](https://user-images.githubusercontent.com/83846226/215345980-37489538-a6ce-482f-ab4a-65295494df56.png)

**Comment by mozman** [comment]
The best solution is a template-based approach that loads the base file first and then adds additional content:

```Python
doc = ezdxf.readfile("MyTemplate.dxf")
msp = doc.modelspace()
# add more stuff to the modelspace
doc.saveas("MyNewFile.dxf")
```

If this doesn't work for you, there is an `Importer` add-on that can import the content of another DXF document:

```Python
from ezdxf.addons import Importer

import_doc = ezdxf.readfile("MyExistingFile.dxf")
doc = ezdxf.new()

importer = Importer(import_doc, doc)
importer.import_modelspace()
importer.finalize()

doc.saveas("MyNewFile.dxf")
```

This solution does not transform the content of the imported document. 
This can be solved by importing the modelspace into a BLOCK and adding a block reference:

```Python

# create the target block in the main document
target_block = doc.blocks.new_anonymous_block()

# import all entities from the modelspace of the imported document into the target block
importer.import_entities(import_doc.modelspace(), target_block)
importer.finalize()

# add a block reference of the target block to the modelspace of the main document
msp.add_blockref(target_block.name, insert=(100, 50))
```

Another solution is to transform the entities of the imported document, this works for simple documents:

```Python
from ezdxf.math import Matrix44

# -x-x-x- snip

m = Matrix44.translate(100, 50, 0)
for entity in import_doc.modelspace():
    entity.transform(m)
```

Please read the [documentation](https://ezdxf.mozman.at/docs/addons/importer.html) of the `Importer` add-on to understand its capabilities and limitations.

EDIT:

Attach an external reference (XREF), but AutoCAD does **not accept** DXF files as XREFs (only DWG is supported which can be created by the `odafc` add-on), BricsCAD and maybe other CAD applications do accept DXF files as XREFs:

```Python
doc.add_xref_def(filename="MyExistingFile.dxf", name="my_xref_block")
msp.add_blockref(name="my_xref_block", insert=(100, 50))
```

Laser cutting software most likely does not support XREFs.
